File: blog_manager/blog_manager/statics/js/apps/libs/models/article.js
import request from '../request.js';
import MarkdownIt from 'markdown-it';

const md = new MarkdownIt();

export default class Model {
  constructor(data) {
    this.data = data || {
      title: '',
      body: '',
      slug: '',
      published: false,
      tags: '',
      date: null,
    };
    this.fetched = false;
    this.resourcesUrl = '/api/articles';
    this.resourceUrl = `${this.resourcesUrl}/${location.pathname.split('/')[2]}`;
  }
  fetch() {
    return request('GET', this.resourceUrl)
      .then(response => {
        this.fetched = true;
        this.data = response;
    });
  }
  create() {
    return request('POST', this.resourcesUrl, this.data)
      .then(response => {
        this.data = response;
    });
  }
  update() {
    return request('PUT', this.resourceUrl, this.data)
      .then(response => {
        this.data = response;
    });
  }
  delete() {
    return request('DELETE', this.resourceUrl)
      .then(response => {
        this.data = null;
      });
  }
  mdBody() {
    return md.render(this.data.body);
  }
  setTag() {
    return v => this.data.tag = v;
  }
}

File: editor/src/handlers/api/article.rs
use rocket_contrib::Json;
use models::{Article, NewArticle, ExistingArticle, Tagging, Tag};
use db::Conn;
use chrono::prelude::*;


#[derive(Serialize)]
struct ArticleWithoutTag {
    id: i32,
    title: String,
    slug: String,
    published: bool,
    created_at: NaiveDateTime,
}

#[get("/articles")]
fn fetch(conn: Conn) -> Json<Vec<ArticleWithoutTag>> {
    let articles: Vec<ArticleWithoutTag> = Article::select_all(&conn)
        .iter()
        .map(|ref a|
            ArticleWithoutTag {
                id: a.clone().id,
                title: a.clone().title.to_owned(),
                slug: a.clone().slug.to_owned(),
                published: a.clone().published,
                created_at: a.clone().created_at,
            }
        )
        .collect();
    Json(articles)
}

#[get("/articles/<id>")]
fn one(id: i32, conn: Conn) -> Json<Article> {
    Json(Article::select(id, &conn))
}

#[get("/articles/<id>/tags")]
fn tags(id: i32, conn: Conn) -> Json<Vec<Tag>> {
    Json(Tagging::tags_by_article(id, &conn))
}

#[post("/articles", format = "application/json", data = "<article>")]
fn create(article: Json<NewArticle>, conn: Conn) -> Json<usize> {
    Json(Article::insert(&article.into_inner(), &conn))
}

#[put("/articles/<id>", format = "application/json", data = "<article>")]
fn update(id: i32, article: Json<ExistingArticle>, conn: Conn) -> Json<usize> {
    Json(Article::update(id, &article.into_inner(), &conn))
}

#[delete("/articles/<id>")]
fn delete(id: i32, conn: Conn) -> Json<usize> {
    Json(Article::delete(id, &conn))
}
